File: output/gemma4-e4b/avatar/Java/Python/atcoder_ABC150_D.py
import sys
import math

def get_gcd(a, b):
    """Calculates the Greatest Common Divisor of a and b."""
    return math.gcd(a, b)

def solve():
    # Read all input tokens from standard input
    data = sys.stdin.read().split()
    if not data:
        return

    # N and M are the first two elements
    N = int(data[0])
    M = int(data[1])
    
    # The remaining N elements are the inputs for A
    inputs_a = [int(x) for x in data[2:2+N]]
    
    # Calculate A: a[i] = input_i / 2 (integer division)
    A = [x // 2 for x in inputs_a]

    # Calculate LCM (lcd)
    lcd = 1
    
    for a_i in A:
        # Calculate GCD(lcd, a_i)
        gcd_val = get_gcd(lcd, a_i)
        
        # Calculate new LCM: lcd * a_i / gcd_val
        # Use integer division //
        lcd = (lcd * a_i) // gcd_val
        
        if lcd > M:
            print(0)
            return

    # Second Check
    for a_i in A:
        # Check if (lcd / a_i) % 2 == 0
        # Since lcd is the LCM, it must be divisible by a_i.
        if (lcd // a_i) % 2 == 0:
            print(0)
            return

    # Final Output: (m / lcd + 1) / 2
    # Use integer division //
    result = (M // lcd + 1) // 2
    print(result)

if __name__ == "__main__":
    solve()

File: output/gemma4-e4b_IO_1/avatar/Python/Java/atcoder_ABC150_D.java
import java.util.*;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.stream.Stream;

public class atcoder_ABC150_D {

    // Helper function for GCD (Greatest Common Divisor)
    private static long gcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return Math.abs(a);
    }

    // Helper function for LCM (Least Common Multiple)
    private static long lcm(long a, long b) {
        if (a == 0 || b == 0) return 0;
        // LCM(a, b) = |a * b| / GCD(a, b)
        return Math.abs(a * b) / gcd(a, b);
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // Read n and num (First line)
        String line1 = br.readLine();
        if (line1 == null || line1.isEmpty()) return;
        String[] firstLine = line1.split(" ");
        long n = Long.parseLong(firstLine[0]);
        long num = Long.parseLong(firstLine[1]);

        // Read the set of numbers (Second line)
        String line2 = br.readLine();
        if (line2 == null || line2.isEmpty()) {
            // Handle case where num set is empty
            System.out.println(0);
            return;
        }
        String[] numStrings = line2.split(" ");
        Set<Long> numSet = new HashSet<>();
        for (String s : numStrings) {
            if (!s.isEmpty()) {
                numSet.add(Long.parseLong(s));
            }
        }

        Set<Integer> twoTimesSet = new HashSet<>();
        boolean shouldBreak = false;

        // Simulate the Python for...else loop structure
        for (long i : numSet) {
            // Use a temporary variable to simulate modification of the loop variable i
            long currentI = i;
            
            for (int j = 1; j < 30; j++) {
                currentI /= 2;
                if (currentI % 2 != 0) {
                    twoTimesSet.add(j);
                    break;
                }
            }
            
            // Check the condition that determines if the loop breaks
            if (twoTimesSet.size() != 1) {
                System.out.println(0);
                shouldBreak = true;
                break;
            }
        }

        // Python's for...else structure: execute this block only if the loop did not break
        if (!shouldBreak) {
            List<Long> numList = new ArrayList<>(numSet);
            
            if (numList.isEmpty()) {
                // If the set was empty, the LCM calculation is undefined/trivial. 
                // Based on the problem context, if numSet is empty, we might assume a default behavior, 
                // but following the logic, if numSet is empty, the loop doesn't run, and we proceed.
                // If numSet is empty, the LCM calculation below will fail if we don't handle it.
                // Assuming the input guarantees at least one element if we reach this point, 
                // or that the calculation should handle it gracefully.
                // Since the original Python code assumes num_list[0] exists if the loop completes, 
                // we assume numSet is non-empty if we reach here.
            } else {
                long lcmVal = numList.get(0);
                for (int i = 1; i < numList.size(); i++) {
                    lcmVal = lcm(lcmVal, numList.get(i));
                }
                
                // Calculate (num - lcm // 2) // (lcm) + 1
                // Note: Integer division // in Python corresponds to standard long division in Java
                long
